packages/components/src/navigation/item.js

```diff
@@ -44,27 +44,27 @@ export default function NavigationItem( {
 
 	return (
 		<ItemUI className={ classes }>
-			<Button href={ href } onClick={ onItemClick } { ...props }>
-				{ title && (
-					<ItemTitleUI
-						className="components-navigation__item-title"
-						variant="body.small"
-						as="span"
-					>
-						{ title }
-					</ItemTitleUI>
-				) }
+			{ children || (
+				<Button href={ href } onClick={ onItemClick } { ...props }>
+					{ title && (
+						<ItemTitleUI
+							className="components-navigation__item-title"
+							variant="body.small"
+							as="span"
+						>
+							{ title }
+						</ItemTitleUI>
+					) }
 
-				{ children }
+					{ badge && (
+						<ItemBadgeUI className="components-navigation__item-badge">
+							{ badge }
+						</ItemBadgeUI>
+					) }
 
-				{ badge && (
-					<ItemBadgeUI className="components-navigation__item-badge">
-						{ badge }
-					</ItemBadgeUI>
-				) }
-
-				{ navigateToMenu && <Icon icon={ chevronRight } /> }
-			</Button>
+					{ navigateToMenu && <Icon icon={ chevronRight } /> }
+				</Button>
+			) }
```

Comment:
This approach has one caveat: custom components would need the whole external state to handle internal navigation/item activation (see the [storybook change](https://github.com/WordPress/gutenberg/pull/25281/files#diff-d0a3484bdc3e3290d7d4cf2205a11b40R71)).

I'd love to keep the component as self-contained as possible by default, leaving the external state as an optional feature, only used if strictly needed.

What if we did some kind of hybrid, and wrap custom components in a div with just the `onClick` handler?
E.g.
```jsx
{ children && <div onClick={ onItemClick }>{ children }</div> }
{ ! children && <Button /> }
```

This would have some pros...
- Clicking the item would activate it without the need of an external state.
- Consumers could define an `onClick` for `NavigationItem` if needed, and a different one for the custom component.
- `navigateToMenu` would work as expected.

and cons...
- There would be no way to prevent `setActiveItem` on click — unless defining the `href` prop... Maybe we could have another prop to optionally disable the activation behaviour? (Sounds overkill 🤔 )
- `navigateToMenu` on custom components would not render the chevron icon. Not a big deal, but it's an inconsistency.

Reply:
>This approach has one caveat: custom components would need the whole external state to handle internal navigation/item activation (see the storybook change).

I do agree that this is a problem and for the WC nav we're building it makes things pretty complicated.  How do we check and control the active state when we have menu items all with various logic to determine active status:
* Traditional wp links
* WooCommerce Admin links
* Other router-based links
* Third party plugin links
* External links

>Clicking the item would activate it without the need of an external state.

I like the simplicity of this and it works for most cases.  One small issue is that clicking does not necessarily guarantee that an item will become active.  One more common case for this might be external links.

I do have an open PR that uses the current route as a single source of truth for determining active state (would need a rebase if we went that direction).  https://github.com/WordPress/gutenberg/pull/24985

**Pros**
* Single source of truth for active state.
* External components can be used without click handlers.
* Works across any type of router.

**Cons**
* Far more opinionated on what is deemed active and requires a route change (or at least a hash change) to trigger changes to the active state of items.
* Lots of testing needed to confirm this works for all use cases.

Reply:
Thanks for the feedback copons!

>What if we did some kind of hybrid, and wrap custom components in a div with just the onClick handler?
E.g.

The linter will complain about a div with an onClick because it will need `role="button"` applied as well, which sort of defeats the purpose because semantically an onClick should only be applied to buttons.

I suppose if you have `setActiveItem` called internally as well as updated from top level props via a change in the url you'd just get an unnecessary render. However, this seems like an indication that something isn't right in the underlying code.

Thanks also @joshuatf 

>One small issue is that clicking does not necessarily guarantee that an item will become active. One more common case for this might be external links.

In the case of external links, no navigation actually occurs in the originating page, so I think this is the expected behaviour.

>I do have an open PR that uses the current route as a single source of truth for determining active state

This is going to be needed by all applications using the Nav component, so I do see the value in bring that in-house. I'm hesitant to advocate for this because it goes beyond the responsibility of the Nav component, which is to render links in a hierarchy. I suppose it's a question of whose responsibility this is. My opinion is why not make a dumb component now and look into adding this later?

Reply:
>In the case of external links, no navigation actually occurs in the originating page, so I think this is the expected behaviour.

In the example above with an internal state, the `onClick` would still fire and would mark this item as "active" which I would consider unexpected behavior since the originating page did not navigate.

>I'm hesitant to advocate for this because it goes beyond the responsibility of the Nav component, which is to render links in a hierarchy. I suppose it's a question of whose responsibility this is. 

I get this and also am a little conflicted about it, but if _every_ consumer is using this logic, I think it should be coupled with the nav.  If we see value in the route detection, but can see use cases for setting active state without it, then I agree that separation is warranted.

>My opinion is why not make a dumb component now and look into adding this later?

I'd argue that this logic is necessary to make the component usable.  While I like the idea of a dumb component, if every consumer is going to write their own route detection logic, why not add it sooner rather than later and make this component more readily usable?

The alternative is just allowing the `isActive` prop to passed to menu items and having consumers pass their own route detection to any registered items.